File: ebpf/c/chanmon.c
#include "vmlinux.h"
#include "maps.h"
#include "goroutine.h"

#include <bpf/bpf_core_read.h>
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>

// read_stack_id reads the stack id from stack trace map.
// 1 on failure
static __always_inline int read_stack_id(struct pt_regs *ctx, int *stack_id) {
    int id = bpf_get_stackid(ctx, &stack_addresses, BPF_F_USER_STACK);
    if (id < 0) {
        bpf_printk("get stack id failed\n");
        return 1;
    }
    *stack_id = id;
    return 0;
}

// func makechan(t *chantype, size int) *hchan
// https://github.com/golang/go/blob/go1.21.4/src/runtime/chan.go#L72
SEC("uretprobe/runtime.makechan")
int runtime_makechan(struct pt_regs *ctx) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    int64_t go_id = 0;
    if (read_goroutine_id(task, &go_id)) {
        return 0;
    }

    int stack_id = 0;
    if (read_stack_id(ctx, &stack_id)) {
        return 0;
    }

    void *sp = (void *)PT_REGS_SP_CORE(ctx);
    int64_t chan_size = 0;
    bpf_core_read_user(&chan_size, sizeof(int64_t), sp + 0x8);

    struct makechan_event_key key = {
        .goroutine_id = go_id,
        .ktime = bpf_ktime_get_ns(),
    };
    struct makechan_event event = {
        .stack_id = stack_id,
        .chan_size = chan_size,
    };
    bpf_map_update_elem(&makechan_events, &key, &event, BPF_ANY);
    return 0;
}

// func chansend1(c *hchan, elem unsafe.Pointer) {
// 	chansend(c, elem, true, getcallerpc())
// }
// https://github.com/golang/go/blob/go1.21.4/src/runtime/chan.go#L144
SEC("uretprobe/runtime.chansend1")
int runtime_chansend1(struct pt_regs *ctx) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    int64_t go_id = 0;
    if (read_goroutine_id(task, &go_id)) {
        return 0;
    }

    int stack_id = 0;
    if (read_stack_id(ctx, &stack_id)) {
        return 0;
    }

    struct chansend_event_key key = {
        .goroutine_id = go_id,
        .ktime = bpf_ktime_get_ns(),
    };
    struct chansend_event event = {
        .stack_id = stack_id,
        .success = (bool)PT_REGS_RC_CORE(ctx),
        .function = chansend1,
    };
    bpf_map_update_elem(&chansend_events, &key, &event, BPF_ANY);
    return 0;
}

// func selectnbsend(c *hchan, elem unsafe.Pointer) (selected bool) {
// 	return chansend(c, elem, false, getcallerpc())
// }
// https://github.com/golang/go/blob/go1.21.4/src/runtime/chan.go#L693
SEC("uretprobe/runtime.selectnbsend")
int runtime_selectnbsend(struct pt_regs *ctx) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    int64_t go_id = 0;
    if (read_goroutine_id(task, &go_id)) {
        return 0;
    }

    int stack_id = 0;
    if (read_stack_id(ctx, &stack_id)) {
        return 0;
    }

    struct chansend_event_key key = {
        .goroutine_id = go_id,
        .ktime = bpf_ktime_get_ns(),
    };
    struct chansend_event event = {
        .stack_id = stack_id,
        .success = (bool)PT_REGS_RC_CORE(ctx),
        .function = selectnbsend,
    };
    bpf_map_update_elem(&chansend_events, &key, &event, BPF_ANY);
    return 0;
}

// func reflect_chansend(c *hchan, elem unsafe.Pointer, nb bool) (selected bool) {
// 	return chansend(c, elem, !nb, getcallerpc())
// }
// https://github.com/golang/go/blob/go1.21.4/src/runtime/chan.go#L718
SEC("uretprobe/runtime.reflect_chansend")
int runtime_reflect_chansend(struct pt_regs *ctx) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    int64_t go_id = 0;
    if (read_goroutine_id(task, &go_id)) {
        return 0;
    }

    int stack_id = 0;
    if (read_stack_id(ctx, &stack_id)) {
        return 0;
    }

    struct chansend_event_key key = {
        .goroutine_id = go_id,
        .ktime = bpf_ktime_get_ns(),
    };
    struct chansend_event event = {
        .stack_id = stack_id,
        .success = (bool)PT_REGS_RC_CORE(ctx),
        .function = reflect_chansend,
    };
    bpf_map_update_elem(&chansend_events, &key, &event, BPF_ANY);
    return 0;
}

// func chanrecv1(c *hchan, elem unsafe.Pointer) {
// 	chanrecv(c, elem, true)
// }
// https://github.com/golang/go/blob/go1.21.4/src/runtime/chan.go#L441
SEC("uretprobe/runtime.chanrecv1")
int runtime_chanrecv1(struct pt_regs *ctx) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    int64_t go_id = 0;
    if (read_goroutine_id(task, &go_id)) {
        return 0;
    }

    int stack_id = 0;
    if (read_stack_id(ctx, &stack_id)) {
        return 0;
    }

    struct chanrecv_event_key key = {
        .goroutine_id = go_id,
        .ktime = bpf_ktime_get_ns(),
    };
    struct chanrecv_event event = {
        .stack_id = stack_id,
        .selected = (bool)PT_REGS_RC_CORE(ctx),
        // We won't use `received` in the Go frontend.
        // See https://github.com/keisku/chanmon/pull/2
        .received = false,
        .function = chanrecv1,
    };
    bpf_map_update_elem(&chanrecv_events, &key, &event, BPF_ANY);
    return 0;
}

// func chanrecv2(c *hchan, elem unsafe.Pointer) (received bool) {
// 	_, received = chanrecv(c, elem, true)
// 	return
// }
// https://github.com/golang/go/blob/go1.21.4/src/runtime/chan.go#L446
SEC("uretprobe/runtime.chanrecv2")
int runtime_chanrecv2(struct pt_regs *ctx) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    int64_t go_id = 0;
    if (read_goroutine_id(task, &go_id)) {
        return 0;
    }

    int stack_id = 0;
    if (read_stack_id(ctx, &stack_id)) {
        return 0;
    }

    struct chanrecv_event_key key = {
        .goroutine_id = go_id,
        .ktime = bpf_ktime_get_ns(),
    };
    struct chanrecv_event event = {
        .stack_id = stack_id,
        // We won't use `selected` in the Go frontend.
        // See https://github.com/keisku/chanmon/pull/2
        .selected = false,
        .received = (bool)PT_REGS_RC_CORE(ctx),
        .function = chanrecv2,
    };
    bpf_map_update_elem(&chanrecv_events, &key, &event, BPF_ANY);
    return 0;
}

// func closechan(c *hchan)
// https://github.com/golang/go/blob/go1.21.4/src/runtime/chan.go#L357
SEC("uretprobe/runtime.closechan")
int runtime_closechan(struct pt_regs *ctx) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    int64_t go_id = 0;
    if (read_goroutine_id(task, &go_id)) {
        return 0;
    }

    int stack_id = 0;
    if (read_stack_id(ctx, &stack_id)) {
        return 0;
    }

    struct closechan_event_key key = {
        .goroutine_id = go_id,
        .ktime = bpf_ktime_get_ns(),
    };
    struct closechan_event event = {
        .stack_id = stack_id,
    };
    bpf_map_update_elem(&closechan_events, &key, &event, BPF_ANY);
    return 0;
}

char LICENSE[] SEC("license") = "GPL";

File: ebpf/chanmon.go
package ebpf

import (
	"context"
	"debug/elf"
	"encoding/binary"
	"fmt"
	"log/slog"
	"strconv"
	"time"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
	"github.com/keisku/chanmon/addr2line"
)

// $BPF_CLANG and $BPF_CFLAGS are set by the Makefile.
//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc $BPF_CLANG -target amd64 -cflags $BPF_CFLAGS bpf ./c/chanmon.c -- -I./c

// maxStackDepth is the max depth of each stack trace to track
// Matches 'MAX_STACK_DEPTH' in eBPF code
const maxStackDepth = 20

var stackFrameSize = (strconv.IntSize / 8)

func Run(ctx context.Context, binPath string, pid int) (context.CancelFunc, error) {
	wrappedCtx, cancel := context.WithCancel(ctx)
	objs := bpfObjects{}
	if err := loadBpfObjects(&objs, nil); err != nil {
		return cancel, err
	}
	elfFile, err := elf.Open(binPath)
	if err != nil {
		return cancel, err
	}
	if err := addr2line.Init(elfFile); err != nil {
		slog.Debug("initialize addr2line", slog.Any("err", err))
	}
	ex, err := link.OpenExecutable(binPath)
	if err != nil {
		return cancel, err
	}
	type uprobeArguments struct {
		symbol string
		prog   *ebpf.Program
		opts   *link.UprobeOptions
		ret    bool
	}
	uprobeArgs := []uprobeArguments{
		{"runtime.makechan", objs.RuntimeMakechan, uprobeOptions(pid), true},
		{"runtime.chansend1", objs.RuntimeChansend1, uprobeOptions(pid), true},
		{"runtime.selectnbsend", objs.RuntimeSelectnbsend, uprobeOptions(pid), true},
		{"runtime.reflect_chansend", objs.RuntimeReflectChansend, uprobeOptions(pid), true},
		{"runtime.chanrecv1", objs.RuntimeChanrecv1, uprobeOptions(pid), true},
		{"runtime.chanrecv2", objs.RuntimeChanrecv2, uprobeOptions(pid), true},
		{"runtime.closechan", objs.RuntimeClosechan, uprobeOptions(pid), true},
		{"runtime.newproc1", objs.RuntimeNewproc1, uprobeOptions(pid), true},
		{"runtime.goexit1", objs.RuntimeGoexit1, uprobeOptions(pid), false},
	}
	uprobeLinks := make([]link.Link, 0, len(uprobeArgs))
	for i := 0; i < len(uprobeArgs); i++ {
		var link link.Link
		var err error
		if uprobeArgs[i].ret {
			link, err = ex.Uretprobe(uprobeArgs[i].symbol, uprobeArgs[i].prog, uprobeArgs[i].opts)
		} else {
			link, err = ex.Uprobe(uprobeArgs[i].symbol, uprobeArgs[i].prog, uprobeArgs[i].opts)
		}
		if err != nil {
			slog.Debug(err.Error())
			continue
		}
		uprobeLinks = append(uprobeLinks, link)
	}
	processes := []func(*bpfObjects) error{
		processMakechanEvents,
		processChansendEvents,
		processChanrecvEvents,
		processClosechanEvents,
		processNewproc1Events,
		processGoexit1Events,
	}
	go func() {
		for {
			select {
			case <-wrappedCtx.Done():
				slog.Debug("eBPF programs stop")
				return
			case <-time.Tick(200 * time.Millisecond):
				for i := 0; i < len(processes); i++ {
					go func(fIdx int) {
						if err := processes[fIdx](&objs); err != nil {
							slog.Warn(err.Error())
						}
					}(i)
				}
			}
		}
	}()
	return func() {
		// Don't use for-range to avoid copying the slice.
		for i := 0; i < len(uprobeLinks); i++ {
			if err := uprobeLinks[i].Close(); err != nil {
				slog.Warn(err.Error())
			}
		}
		if err := objs.Close(); err != nil {
			slog.Warn("Failed to close bpf objects: %s", err)
		}
		cancel()
	}, nil
}

func processMakechanEvents(objs *bpfObjects) error {
	var key bpfMakechanEventKey
	var value bpfMakechanEvent
	var keysToDelete []bpfMakechanEventKey
	return processEvents(
		objs.StackAddresses,
		objs.MakechanEvents,
		func(mapIter *ebpf.MapIterator, stackIdSet map[int32]struct{}) (any, int) {
			for mapIter.Next(&key, &value) {
				stack, err := extractStack(objs, value.StackId)
				if err != nil {
					slog.Warn(err.Error())
					continue
				}
				stackIdSet[value.StackId] = struct{}{}
				keysToDelete = append(keysToDelete, key)
				slog.Info("runtime.makechan",
					slog.Int64("goroutine_id", int64(key.GoroutineId)),
					slog.Int64("stack_id", int64(value.StackId)),
					slog.Int64("chan_size", int64(value.ChanSize)),
					stackToLogAttr(stack),
				)
			}
			return keysToDelete, len(keysToDelete)
		},
	)
}

func processChansendEvents(objs *bpfObjects) error {
	var key bpfChansendEventKey
	var value bpfChansendEvent
	var keysToDelete []bpfChansendEventKey
	translation := [...]string{
		"unknown",
		"runtime.chansend1",
		"runtime.selectnbsend",
		"runtime.reflect_chansend",
	}
	return processEvents(
		objs.StackAddresses,
		objs.ChansendEvents,
		func(mapIter *ebpf.MapIterator, stackIdSet map[int32]struct{}) (any, int) {
			for mapIter.Next(&key, &value) {
				if value.Function == 0 { // unknown
					slog.Error("unreachable")
					continue
				}
				stack, err := extractStack(objs, value.StackId)
				if err != nil {
					slog.Warn(err.Error())
					continue
				}
				stackIdSet[value.StackId] = struct{}{}
				keysToDelete = append(keysToDelete, key)
				slog.Info("runtime.chansend",
					slog.Int64("goroutine_id", int64(key.GoroutineId)),
					slog.Int64("stack_id", int64(value.StackId)),
					slog.Bool("success", value.Success),
					slog.String("function", translation[value.Function]),
					stackToLogAttr(stack),
				)
			}
			return keysToDelete, len(keysToDelete)
		},
	)
}

func processChanrecvEvents(objs *bpfObjects) error {
	var key bpfChanrecvEventKey
	var value bpfChanrecvEvent
	var keysToDelete []bpfChanrecvEventKey
	return processEvents(
		objs.StackAddresses,
		objs.ChanrecvEvents,
		func(mapIter *ebpf.MapIterator, stackIdSet map[int32]struct{}) (any, int) {
			for mapIter.Next(&key, &value) {
				if value.Function == 0 { // unknown
					slog.Error("unreachable")
					continue
				}
				stack, err := extractStack(objs, value.StackId)
				if err != nil {
					slog.Warn(err.Error())
					continue
				}
				stackIdSet[value.StackId] = struct{}{}
				keysToDelete = append(keysToDelete, key)
				attrs := []any{
					slog.Int64("goroutine_id", int64(key.GoroutineId)),
					slog.Int64("stack_id", int64(value.StackId)),
					stackToLogAttr(stack),
				}
				// As of Go version 1.21.5, there is no mechanism to access `selected` and `received`,
				// which are the first and second return values of `runtime.chanrecv`, respectively.
				// They are stored in the rax and rbx registers of the amd64 architecture.
				// Upon examining the assembly code of `runtime.chanrecv` through gdb,
				// it's evident that the return values aren't stored at the stack pointer,
				// thereby rendering it impossible to retrieve `selected` and `received` both.
				// See https://github.com/keisku/chanmon/pull/2
				switch value.Function {
				case 1:
					attrs = append(
						attrs,
						slog.Bool("selected", value.Selected),
						slog.String("function", "runtime.chanrecv1"),
					)
				case 2:
					attrs = append(
						attrs,
						slog.Bool("received", value.Received),
						slog.String("function", "runtime.chanrecv2"),
					)
				}
				slog.Info("runtime.chanrecv", attrs...)
			}
			return keysToDelete, len(keysToDelete)
		},
	)
}

func processClosechanEvents(objs *bpfObjects) error {
	var key bpfClosechanEventKey
	var value bpfClosechanEvent
	var keysToDelete []bpfClosechanEventKey

	return processEvents(
		objs.StackAddresses,
		objs.ClosechanEvents,
		func(mapIter *ebpf.MapIterator, stackIdSet map[int32]struct{}) (any, int) {
			for mapIter.Next(&key, &value) {
				stack, err := extractStack(objs, value.StackId)
				if err != nil {
					slog.Warn(err.Error())
					continue
				}
				stackIdSet[value.StackId] = struct{}{}
				keysToDelete = append(keysToDelete, key)
				slog.Info("runtime.closechan",
					slog.Int64("goroutine_id", int64(key.GoroutineId)),
					slog.Int64("stack_id", int64(value.StackId)),
					stackToLogAttr(stack),
				)
			}
			return keysToDelete, len(keysToDelete)
		},
	)
}

func processNewproc1Events(objs *bpfObjects) error {
	var key bpfNewproc1EventKey
	var value bpfNewproc1Event
	var keysToDelete []bpfNewproc1EventKey

	return processEvents(
		objs.StackAddresses,
		objs.Newproc1Events,
		func(mapIter *ebpf.MapIterator, stackIdSet map[int32]struct{}) (any, int) {
			for mapIter.Next(&key, &value) {
				stack, err := extractStack(objs, value.StackId)
				if err != nil {
					slog.Warn(err.Error())
					continue
				}
				stackIdSet[value.StackId] = struct{}{}
				keysToDelete = append(keysToDelete, key)
				slog.Info("runtime.newproc1",
					slog.Int64("goroutine_id", int64(key.GoroutineId)),
					slog.Int64("stack_id", int64(value.StackId)),
					stackToLogAttr(stack),
				)
			}
			return keysToDelete, len(keysToDelete)
		},
	)
}

func processGoexit1Events(objs *bpfObjects) error {
	var key bpfGoexit1EventKey
	var value bpfGoexit1Event
	var keysToDelete []bpfGoexit1EventKey

	return processEvents(
		objs.StackAddresses,
		objs.Goexit1Events,
		func(mapIter *ebpf.MapIterator, stackIdSet map[int32]struct{}) (any, int) {
			for mapIter.Next(&key, &value) {
				stack, err := extractStack(objs, value.StackId)
				if err != nil {
					slog.Warn(err.Error())
					continue
				}
				stackIdSet[value.StackId] = struct{}{}
				keysToDelete = append(keysToDelete, key)
				slog.Info("runtime.goexit1",
					slog.Int64("goroutine_id", int64(key.GoroutineId)),
					slog.Int64("stack_id", int64(value.StackId)),
					stackToLogAttr(stack),
				)
			}
			return keysToDelete, len(keysToDelete)
		},
	)
}

func processEvents(
	stackAddrs, eventMap *ebpf.Map,
	// stackIdSet is the set of stack_id to delete later.
	// keysToDelete is the slice of eBPF map keys to delete later.
	// keyLength holds the count of keys in keysToDelete to determine if BatchDelete is required.
	processMap func(iter *ebpf.MapIterator, stackIdSet map[int32]struct{}) (keysToDelete any, keyLength int),
) error {
	stackIdSetToDelete := make(map[int32]struct{})
	mapIter := eventMap.Iterate()
	keysToDelete, keyLength := processMap(mapIter, stackIdSetToDelete)
	if err := mapIter.Err(); err != nil {
		return fmt.Errorf("failed to iterate eBPF map: %w", err)
	}
	if 0 < keyLength {
		if n, err := eventMap.BatchDelete(keysToDelete, nil); err == nil {
			slog.Debug("Deleted eBPF map", slog.Int("deleted", n), slog.Int("expected", keyLength))
		} else {
			slog.Warn("Failed to delete eBPF map", slog.String("error", err.Error()))
		}
	}
	// Don't use BatchDelete for stack addresses because the opration is not supported.
	// If we do it, we will see "batch delete: not supported" error.
	for stackId := range stackIdSetToDelete {
		if err := stackAddrs.Delete(stackId); err != nil {
			slog.Warn("Failed to delete stack_addresses", slog.String("error", err.Error()))
			continue
		}
		slog.Debug("Deleted stack address map", slog.Int("stack_id", int(stackId)))
	}
	return nil
}

func extractStack(objs *bpfObjects, stackId int32) ([]string, error) {
	stack := make([]string, maxStackDepth)
	stackBytes, err := objs.StackAddresses.LookupBytes(stackId)
	if err != nil {
		return nil, fmt.Errorf("failed to lookup stack address: %w", err)
	}
	stackCounter := 0
	for i := 0; i < len(stackBytes); i += stackFrameSize {
		stackBytes[stackCounter] = 0
		stackAddr := binary.LittleEndian.Uint64(stackBytes[i : i+stackFrameSize])
		if stackAddr == 0 {
			break
		}
		if line := addr2line.Do(stackAddr); line == "" {
			stack[stackCounter] = fmt.Sprintf("%x", stackAddr)
		} else {
			stack[stackCounter] = line
		}
		stackCounter++
	}
	return stack[0:stackCounter], nil
}

func stackToLogAttr(stack []string) slog.Attr {
	attrs := make([]slog.Attr, len(stack))
	for i, s := range stack {
		attrs[i] = slog.String(fmt.Sprintf("%d", i), s)
	}
	return slog.Attr{
		Key:   "stack",
		Value: slog.GroupValue(attrs...),
	}
}

func uprobeOptions(pid int) *link.UprobeOptions {
	if 0 < pid {
		return &link.UprobeOptions{PID: pid}
	}
	return nil
}
